### PR TITLE
Mistral Support

### PR DESCRIPTION
**Pull request by YanSte**
This pull request introduces integration with the Mistral model for Pydantic AI. 

The MistralModel class leverages the Mistral Python client to interact with the Mistral API, enabling both streaming and non-streaming requests and Structured Responses.

This integration supports various modes of operation, including function calling, JSON mode, and stream mode, based on the presence of function and result tools.

---

Note: Mistral does not support Streaming on Function Calling or Structured Responses. 
Even when using stream_async, the behavior is not truly streaming. After discussions on the Mistral Discord, I use json_mode to stream only the structured output.

---
[Sample Notebook](https://github.com/YanSte/pydantic-ai/blob/nootbook/pydantic_ai_examples/mistral.ipynb)

---

Test: >> Working on it. <<

- [x] Tests

### CODE REVIEW

File: pydantic_ai_examples/mistral.ipynb
Comment (samuelcolvin) [inline pydantic_ai_examples/mistral.ipynb:1]: Please let's not use jupyter notebooks, I'm [not a fan](https://github.com/samuelcolvin/notbook/blob/master/README.md).

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+            self.client = Mistral(
+                api_key=api_key, async_client=cached_async_http_client()
+            )
+
+    async def agent_model(
+        self,
+        function_tools: Mapping[str, AbstractToolDefinition],
+        allow_text_result: bool,
+        result_tools: Sequence[AbstractToolDefinition] | None,
+    ) -> AgentModel:
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: I'm afraid you'll need to update this since I merged #157, it shouldn't be too hard.

**Review by samuelcolvin — COMMENTED** [review]
otherwise looks like a good start.

**Comment by YanSte** [comment]
@samuelcolvin I'm having some trouble with the Stream Structured part. Could you take a look at it?

**Comment by YanSte** [comment]
Good I solved it. @samuelcolvin

**Comment by samuelcolvin** [comment]
@YanSte `BaseModel` isn't yet supported by partial validation, this is explained in data at https://docs.pydantic.dev/latest/concepts/experimental/#partial-validation.

Use a `TypedDict` and it should work.

**Comment by YanSte** [comment]
> orted by partial validation, this

Yes, I remember now, I overlooked this part. 
To ensure we don't miss it again, maybe add an update to documentation to include this information.

**Comment by YanSte** [comment]
Hi @samuelcolvin,

I need some assistance in writing the test to simulate a call from Mistral. Could you please help me with this?

Thank you.

File: tests/models/test_mistral.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+from functools import cached_property
+from typing import Any, cast
+
+import pytest
+from inline_snapshot import snapshot
+
+from pydantic_ai import _utils
+from pydantic_ai.agent import Agent
+from pydantic_ai.messages import ModelTextResponse, UserPrompt
+from pydantic_ai.models.mistral import MistralModel
```
Comment (samuelcolvin) [inline tests/models/test_mistral.py]: You need to use `try_import()` and the skip tests if mistral isn't installed, like this:

https://github.com/pydantic/pydantic-ai/blob/4edb122d1358bc7572857a75a92748015b749010/tests/models/test_openai.py#L29-L48

File: pyproject.toml [hunk 3]
```diff
@@ -36,7 +36,7 @@ classifiers = [
     "Framework :: Pytest",
 ]
 requires-python = ">=3.9"
-dependencies = ["pydantic-ai-slim[openai,vertexai,groq]==0.0.12"]
+dependencies = ["pydantic-ai-slim[openai,vertexai,groq,mistral]==0.0.13"]
```
Comment (samuelcolvin) [inline pyproject.toml]:
```suggestion
dependencies = ["pydantic-ai-slim[openai,vertexai,groq,mistral]==0.0.12"]
```
we're still on `v0.0.12`

**Review by samuelcolvin — COMMENTED** [review]
I haven't looked at all the code yet, but this is why tests are failing.

File: tests/models/test_mistral.py (hunk 2 above)
Comment (YanSte) [reply tests/models/test_mistral.py]: Thanks for your return, I will have a look.

Comment (YanSte) [reply tests/models/test_mistral.py]: Good 🎉 Thanks @samuelcolvin .

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+    [API Documentation](https://docs.mistral.ai/)
+
+    """
+
+    model_name: MistralModelName | str
+    client: Mistral = field(repr=False)
+
+    def __init__(
+        self,
+        model_name: MistralModelName,
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: The types for `model_name` should be the same.

If you want to support other model names, we can use the same pattern as ollama and anthropic, shown here

https://github.com/pydantic/pydantic-ai/blob/01565f59b77c7d0b9343a011edacefe2136993f9/pydantic_ai_slim/pydantic_ai/models/anthropic.py#L53-L66

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+    from mistralai.models.assistantmessage import AssistantMessage as MistralAssistantMessage
+    from mistralai.models.function import Function as MistralFunction
+    from mistralai.models.toolmessage import ToolMessage as MistralToolMessage
+    from mistralai.models.usermessage import UserMessage as MistralUserMessage
+    from mistralai.types.basemodel import Unset as MistralUnset
+    from mistralai.utils.eventstreaming import EventStreamAsync as MistralEventStreamAsync
+except ImportError as e:
+    raise ImportError(
+        'Please install `mistral` to use the Mistral model, '
+        "you can use the `mistral` optional group — `pip install 'pydantic-ai[mistral]'`"
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
        "you can use the `mistral` optional group — `pip install 'pydantic-ai-slim[mistral]'`"
```

I guess we'll need to update other places too.

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+        Args:
+            model_name: The name of the model to use.
+            api_key: The API key to use for authentication,
+            client: An existing `Mistral` client to use, if provided, `api_key` and `http_client` must be `None`.
+            http_client: An existing `httpx.AsyncClient` to use for making HTTP requests.
+        """
+        self.model_name = model_name
+
+        if client is not None:
+            assert http_client is None, 'Cannot provide both `mistral_client` and `http_client`'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py:117]: and check `api_key is None`.

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+    def name(self) -> str:
+        return f'mistral:{self.model_name}'
+
+
+@dataclass
+class MistralAgentModel(AgentModel):
+    """Implementation of `AgentModel` for Mistral models."""
+
+    client: Mistral
+    model_name: str
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py:149]: should have same type as `model_name` above.

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+
+    async def _completions_create(
+        self,
+        messages: list[Message],
+    ) -> MistralChatCompletionResponse:
+        """Make a non-streaming request to the model."""
+        mistral_messages = [self._map_message(m) for m in messages]
+        # Note: Function calling Mode
+        # "auto": default mode. Model decides if it uses the tool or not.
+        # "any": forces tool use. WARN: If use 'any' it will be infinity loop with Model.
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: I don't think this warning is true - we use tools for structured return types.

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+        # "none": prevents tool use.
+        tool_choice: Literal['none', 'any', 'auto'] | None = None
+        if not self.function_tools and not self.result_tools:
+            tool_choice = 'none'
+        else:
+            tool_choice = 'auto'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: I think we should use `any` if `allow_text_result is False`

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+
+        response = await self.client.chat.complete_async(
+            model=str(self.model_name),
+            messages=mistral_messages,
+            n=1,
+            tools=self._map_function_and_result_tools_definition(),
+            tool_choice=tool_choice,
+            stream=False,
+        )
+        assert response, 'A unexpected empty response.'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
        assert response, 'A unexpected empty response from Mistral.'
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+
+    async def _stream_completions_create(
+        self,
+        messages: list[Message],
+    ) -> MistralEventStreamAsync[MistralCompletionEvent]:
+        """Create a streaming completion request to the Mistral model."""
+        response: MistralEventStreamAsync[MistralCompletionEvent] | None = None
+        mistral_messages = [self._map_message(m) for m in messages]
+
+        if self.result_tools and self.function_tools or self.function_tools:
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py:203]:
```suggestion
        if self.result_tools and self.function_tools:
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+                messages=mistral_messages,
+                response_format={'type': 'json_object'},
+            )
+
+        else:
+            # Stream Mode
+            response = await self.client.chat.stream_async(
+                model=str(self.model_name), messages=mistral_messages, stream=True, n=1
+            )
+        assert response
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
        assert response, 'A unexpected empty response from Mistral.'
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+        return response
+
+    def _map_function_and_result_tools_definition(self) -> list[MistralTool] | None:
+        """Map function and result tools to MistralTool format.
+
+        Returns None if both function_tools and result_tools are empty.
+        """
+        tools = []
+
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+@dataclass
+class MistralAgentModel(AgentModel):
+    """Implementation of `AgentModel` for Mistral models."""
+
+    client: Mistral
+    model_name: str
+    allow_text_result: bool
+    function_tools: list[ToolDefinition] | None
+    result_tools: list[ToolDefinition] | None
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
    function_tools: list[ToolDefinition]
    result_tools: list[ToolDefinition]
```
If you simplify the types here, there will be less logic above, and less logic within `MistralAgentModel` 🎉

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+
+        Returns None if both function_tools and result_tools are empty.
+        """
+        tools = []
+
+        all_tools: list[ToolDefinition] = []
+        if self.function_tools:
+            all_tools.extend(self.function_tools)
+        if self.result_tools:
+            all_tools.extend(self.result_tools)
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: with the simplification of types above, this becomes just 
```suggestion
        all_tools: list[ToolDefinition] = self.function_tools + self.result_tools
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+                function=MistralFunction(name=r.name, parameters=r.parameters_json_schema, description=r.description)
+            )
+            for r in all_tools
+        ]
+        return tools if tools else None
+
+    @staticmethod
+    def _process_response(response: MistralChatCompletionResponse) -> ModelAnyResponse:
+        """Process a non-streamed response, and prepare a message to return."""
+        timestamp: datetime
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
```
not needed

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 17]
```diff
@@ -0,0 +1,575 @@
...
+
+        tools_calls = choice.message.tool_calls
+        if tools_calls is not None and isinstance(tools_calls, list):
+            return ModelStructuredResponse(
+                [_map_mistral_to_pydantic_tool_call(tool_call) for tool_call in tools_calls],
+                timestamp=timestamp,
+            )
+        else:
+            content = choice.message.content
+            assert content, f'Unexpected null content is assitant msg: {choice.message}'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
            assert content, f'Unexpected null content in assistant msg: {choice.message}'
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 18]
```diff
@@ -0,0 +1,575 @@
...
+            assert not isinstance(
+                content, list
+            ), f'Unexpected ContentChunk from stream, need to be response not stream: {content}'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
            ), f'Unexpected ContentChunk from stream, got {type(content)}, expected list'
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+                        timestamp,
+                        start_cost,
+                    )
+
+    @staticmethod
+    def _map_message(message: Message) -> MistralMessages:
+        """Just maps a `pydantic_ai.Message` to a `Mistral Message`."""
+        if message.role == 'system':
+            # SystemPrompt ->
+            return MistralAssistantMessage(content=message.content)
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]: this looks wrong, it shouldn't be an assistant message, if theres no specific system prompt type, I would use `MistralUserMessage`.

Unless I'm missing something?

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 20]
```diff
@@ -0,0 +1,575 @@
...
+    def timestamp(self) -> datetime:
+        return self._timestamp
+
+
+def _generate_json_simple_schema(schema: dict[str, Any]) -> Any:
+    """Generates a JSON example from a JSON schema."""
+    if schema.get('type') == 'object':
+        example: dict[str, Any] = {}
+        if 'properties' in schema:
+            for key, value in schema['properties'].items():
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
        if properties := schema.get('properties'):
            for key, value in properties.items():
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 21]
```diff
@@ -0,0 +1,575 @@
...
+                example[key] = _generate_json_simple_schema(value)
+        return example
+
+    if schema.get('type') == 'array':
+        if 'items' in schema:
+            return [_generate_json_simple_schema(schema['items'])]
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
    if schema.get('type') == 'array':
        if items := schema.get('items')
            return [_generate_json_simple_schema(items)]
```

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 22]
```diff
@@ -0,0 +1,575 @@
...
+    if schema.get('type') == 'boolean':
+        return 'Boolean value'
+
+    if schema.get('type') == 'null':
+        return 'null value'
+
+    return None
+
+
+def _generate_jsom_simple_schemas(schemas: list[dict[str, Any]]) -> list[dict[str, Any]]:
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
def _generate_json_simple_schemas(schemas: list[dict[str, Any]]) -> list[dict[str, Any]]:
```
!!!

Comment (samuelcolvin) [reply pydantic_ai_slim/pydantic_ai/models/mistral.py]: do you even need this method? surely

```py
schema = [_generate_json_simple_schema(s) for s in  parameters_json_schemas]
```

would be clearer

File: pydantic_ai_slim/pydantic_ai/models/mistral.py [hunk 23]
```diff
@@ -0,0 +1,575 @@
...
+    if isinstance(delta_content, list) and isinstance(delta_content[0], MistralTextChunk):
+        content = delta_content[0].text
+    elif isinstance(delta_content, str):
+        content = delta_content
+    elif isinstance(delta_content, MistralUnset) or delta_content is None:
+        content = None
+    else:
+        assert (
+            False
+        ), f'Other type of instance, Will manage in the futur (Image, Reference), delta content:{delta_content}'
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/models/mistral.py]:
```suggestion
        ), f'Other data types like (Image, Reference) are not yet supported,  got {type(delta_content)}'
```

File: pyproject.toml [hunk 24]
```diff
@@ -36,7 +36,8 @@ classifiers = [
     "Framework :: Pytest",
 ]
 requires-python = ">=3.9"
-dependencies = ["pydantic-ai-slim[openai,vertexai,groq,anthropic]==0.0.12"]
+
+dependencies = ["pydantic-ai-slim[openai,vertexai,groq,anthropic, mistral]==0.0.12"]
```
Comment (samuelcolvin) [inline pyproject.toml]:
```suggestion
dependencies = ["pydantic-ai-slim[openai,vertexai,groq,anthropic,mistral]==0.0.12"]
```

**Review by samuelcolvin — COMMENTED** [review]
This is looking good, thank so much for your work on this

Coverage needs to be much better, ideally <10 lines not covered

image

**Comment by YanSte** [comment]
Thanks @samuelcolvin 

I will finish the tests (I'm still working on it).

Improving the streaming json mode.

And ping you back.

**Comment by dmontagu** [comment]
I've got this merged with main, just let me merge it with your latest pushes

**Comment by dmontagu** [comment]
You should definitely check the changes I made, in particular commenting out `# created=1704067200,  # 2024-01-01` in the `tests.models.test_mistral.chunk` function. I dropped it to get the tests passing but if you want to confirm the timestamp setting is correct you might want to tweak that.

But I figured I would probably be faster to set up the model settings stuff and deal with the refactor to the new unified `ModelResponse` type, and that you'd appreciate the help. If you don't feel free to revert my push!

**Pull request merged**